File: src/scripts/analyze-claude-lines.py
#!/usr/bin/env python3
"""
Analyze Claude Code contributions across commits and map them to final diff lines.

This script is used by GitHub Actions to preserve Claude Code tracking data
when PRs are squashed, ensuring accurate attribution in the final commit.
"""

import sys
import re
import subprocess
from collections import defaultdict
from typing import Dict, Set, List, Tuple


def parse_claude_data(filename: str) -> Dict[str, Set[int]]:
    """Parse the collected Claude commit data from the pre-merge workflow."""
    claude_files = defaultdict(set)
    
    try:
        with open(filename, 'r') as f:
            for line in f:
                parts = line.strip().split('|')
                if len(parts) == 3:
                    commit_hash, filepath, ranges = parts
                    
                    # Parse ranges like "10-20,25-30"
                    for range_str in ranges.split(','):
                        range_str = range_str.strip()
                        if '-' in range_str:
                            try:
                                start, end = map(int, range_str.split('-'))
                                claude_files[filepath].update(range(start, end + 1))
                            except ValueError:
                                print(f"Warning: Could not parse range '{range_str}' in file {filepath}")
                        else:
                            try:
                                claude_files[filepath].add(int(range_str))
                            except ValueError:
                                print(f"Warning: Could not parse line number '{range_str}' in file {filepath}")
    except FileNotFoundError:
        print(f"Error: Could not find Claude data file: {filename}")
        return defaultdict(set)
    except Exception as e:
        print(f"Error parsing Claude data: {e}")
        return defaultdict(set)
    
    return claude_files


def get_final_diff_lines(base_commit: str, latest_commit: str) -> Dict[str, Set[int]]:
    """Get the actual lines that were added/modified in the final diff."""
    try:
        result = subprocess.run([
            'git', 'diff', '--unified=0', f'{base_commit}..{latest_commit}'
        ], capture_output=True, text=True, check=True)
    except subprocess.CalledProcessError as e:
        print(f"Error running git diff: {e}")
        return defaultdict(set)
    
    final_lines = defaultdict(set)
    current_file = None
    
    for line in result.stdout.split('\n'):
        if line.startswith('+++'):
            # Extract filename, removing '+++ b/' prefix
            current_file = line[6:] if len(line) > 6 else None
        elif line.startswith('@@') and current_file:
            # Parse hunk header like @@ -1,4 +10,8 @@
            match = re.search(r'\+(\d+)(?:,(\d+))?', line)
            if match:
                try:
                    start_line = int(match.group(1))
                    count = int(match.group(2)) if match.group(2) else 1
                    
                    # Mark these lines as part of the final diff
                    for i in range(start_line, start_line + count):
                        final_lines[current_file].add(i)
                except ValueError:
                    print(f"Warning: Could not parse hunk header: {line}")
    
    return final_lines


def map_claude_to_final(claude_files: Dict[str, Set[int]], final_lines: Dict[str, Set[int]]) -> Dict[str, Set[int]]:
    """
    Map Claude's original line contributions to final line numbers.
    
    This is a simplified approach that assumes if Claude touched a file
    and the file has changes in the final diff, then Claude contributed
    to those changes. A more sophisticated approach would need to trace
    line-by-line changes through git history.
    """
    final_claude_lines = defaultdict(set)
    
    for filepath in claude_files:
        if filepath in final_lines:
            # Simple mapping: if Claude touched the file and it has final changes,
            # assume Claude contributed to the final changes
            final_claude_lines[filepath] = final_lines[filepath]
    
    return final_claude_lines


def convert_lines_to_ranges(lines: List[int]) -> str:
    """Convert a list of line numbers to compact range notation like '10-20,25-30'."""
    if not lines:
        return ""
    
    sorted_lines = sorted(lines)
    ranges = []
    start = sorted_lines[0]
    end = sorted_lines[0]
    
    for line in sorted_lines[1:]:
        if line == end + 1:
            end = line
        else:
            # Close current range
            if start == end:
                ranges.append(str(start))
            else:
                ranges.append(f"{start}-{end}")
            start = end = line
    
    # Add the final range
    if start == end:
        ranges.append(str(start))
    else:
        ranges.append(f"{start}-{end}")
    
    return ",".join(ranges)


def generate_claude_note(final_claude_lines: Dict[str, Set[int]]) -> str:
    """Generate the final claude-was-here note in the standard format."""
    lines = ['claude-was-here', 'version: 1.0']
    
    if final_claude_lines:
        # Calculate max path length for alignment
        max_length = max(len(path) for path in final_claude_lines.keys())
        
        # Add each file with its ranges
        for filepath in sorted(final_claude_lines.keys()):
            line_set = final_claude_lines[filepath]
            ranges_str = convert_lines_to_ranges(list(line_set))
            if ranges_str:
                padded_path = f"{filepath}:".ljust(max_length + 2)
                lines.append(f"{padded_path} {ranges_str}")
    
    return '\n'.join(lines)


def main():
    """Main execution function."""
    if len(sys.argv) < 4:
        print("Usage: python analyze_claude_lines.py <claude_data_file> <base_commit> <latest_commit>")
        print("")
        print("This script analyzes Claude Code contributions across commits and maps them")
        print("to the final diff, generating accurate attribution for squashed commits.")
        sys.exit(1)
    
    claude_data_file = sys.argv[1]
    base_commit = sys.argv[2]
    latest_commit = sys.argv[3]
    
    # Parse Claude's original contributions
    claude_files = parse_claude_data(claude_data_file)
    print(f"Found Claude contributions in {len(claude_files)} files", file=sys.stderr)
    
    # Get final diff lines
    final_lines = get_final_diff_lines(base_commit, latest_commit)
    print(f"Final diff affects {len(final_lines)} files", file=sys.stderr)
    
    # Map Claude contributions to final lines
    final_claude_lines = map_claude_to_final(claude_files, final_lines)
    
    # Generate and output the final note
    note_content = generate_claude_note(final_claude_lines)
    print(note_content)


if __name__ == "__main__":
    main()

File: src/scripts/analyze-claude-lines.ts
#!/usr/bin/env -S bun run
/**
 * Analyze Claude Code contributions across commits and map them to final diff lines.
 * 
 * This script is used by GitHub Actions to preserve Claude Code tracking data
 * when PRs are squashed, ensuring accurate attribution in the final commit.
 * 
 * NOTE: This script contains embedded logic from the claude-was-here shared module
 * to avoid import dependencies when installed in other repositories.
 */

import { spawn } from 'child_process';
import { readFile } from 'fs/promises';
import { join } from 'path';

interface ClaudeContribution {
  commitHash: string;
  filepath: string;
  ranges: string;
}

interface ContentSignatures {
  hashes: Set<string>;
}

interface ClaudeLineMapping {
  [filepath: string]: Set<number>;
}

/**
 * Execute a git command and return the result
 */
const execGitCommand = (args: string[], cwd: string): Promise<{ stdout: string; stderr: string; code: number }> => {
  return new Promise((resolve) => {
    const proc = spawn('git', args, { 
      cwd, 
      stdio: ['ignore', 'pipe', 'pipe']
    });
    
    let stdout = '';
    let stderr = '';
    
    proc.stdout.on('data', (data) => stdout += data.toString());
    proc.stderr.on('data', (data) => stderr += data.toString());
    
    proc.on('close', (code) => {
      resolve({ stdout: stdout.trim(), stderr: stderr.trim(), code: code || 0 });
    });
  });
};

/**
 * Collect Claude notes from all commits in a range
 */
const collectClaudeNotesFromCommits = async (
  testDir: string, 
  baseCommit: string, 
  headCommit: string
): Promise<{ contributions: ClaudeContribution[], contentSignatures: ContentSignatures }> => {
  const commitsResult = await execGitCommand(['log', '--format=%H', `${baseCommit}..${headCommit}`], testDir);
  const commits = commitsResult.stdout.split('\n').filter(hash => hash.trim());
  
  const contributions: ClaudeContribution[] = [];
  const contentSignatures: ContentSignatures = { hashes: new Set() };
  
  for (const commitHash of commits) {
    const notesResult = await execGitCommand(['notes', 'show', commitHash], testDir);
    if (notesResult.code === 0) {
      const noteLines = notesResult.stdout.split('\n');
      
      for (const line of noteLines) {
        if (line === 'claude-was-here' || line.startsWith('version:')) {
          continue;
        }
        
        // Handle content-signatures line
        if (line.startsWith('content-signatures:')) {
          const hashesStr = line.substring('content-signatures:'.length).trim();
          if (hashesStr) {
            const hashes = hashesStr.split(',').map(h => h.trim()).filter(h => h);
            hashes.forEach(hash => contentSignatures.hashes.add(hash));
          }
          continue;
        }
        
        const match = line.match(/^([^:]+):\s+(.+)$/);
        if (match) {
          const filepath = match[1].trim();
          const ranges = match[2].trim();
          contributions.push({ commitHash, filepath, ranges });
        }
      }
    }
  }
  
  return { contributions, contentSignatures };
};

// Note: Removed getFinalDiffLines - we now use git notes data directly

/**
 * Consolidate Claude contributions from git notes into final line mappings
 * This directly uses the line authorship data from git notes without pattern matching
 */
const consolidateClaudeContributions = async (
  testDir: string,
  contributions: ClaudeContribution[]
): Promise<ClaudeLineMapping> => {
  const finalClaudeLines: ClaudeLineMapping = {};
  
  // Parse ranges from string like "10-20,25-30" into line numbers
  const parseRanges = (ranges: string): number[] => {
    const lines: number[] = [];
    
    for (const rangeStr of ranges.split(',')) {
      const trimmed = rangeStr.trim();
      if (trimmed.includes('-')) {
        const [start, end] = trimmed.split('-').map(n => parseInt(n));
        for (let i = start; i <= end; i++) {
          lines.push(i);
        }
      } else {
        lines.push(parseInt(trimmed));
      }
    }
    
    return lines;
  };
  
  // Consolidate all Claude contributions by file
  for (const contribution of contributions) {
    const { filepath, ranges } = contribution;
    
    // Initialize file set if not exists
    if (!finalClaudeLines[filepath]) {
      finalClaudeLines[filepath] = new Set();
    }
    
    // Parse the ranges and add all Claude-authored lines
    const lines = parseRanges(ranges);
    for (const lineNum of lines) {
      finalClaudeLines[filepath].add(lineNum);
    }
  }
  
  // Filter out files that don't exist and validate line numbers are within bounds
  const existingFiles: ClaudeLineMapping = {};
  for (const [filepath, lineSet] of Object.entries(finalClaudeLines)) {
    try {
      const fileContent = await readFile(join(testDir, filepath), 'utf-8');
      const totalLines = fileContent.split('\n').length;
      
      // Filter out line numbers that exceed the file's actual line count
      const validLines = new Set<number>();
      for (const lineNum of lineSet) {
        if (lineNum >= 1 && lineNum <= totalLines) {
          validLines.add(lineNum);
        }
      }
      
      // Only include the file if it has valid Claude lines
      if (validLines.size > 0) {
        existingFiles[filepath] = validLines;
      }
    } catch (error) {
      // File doesn't exist in final version, skip it
      continue;
    }
  }
  
  return existingFiles;
};

// Note: getFinalDiffLines function is no longer needed since we use git notes data directly

/**
 * Convert line numbers to compact range notation
 */
const convertLinesToRanges = (lines: number[]): string => {
  if (lines.length === 0) return '';
  
  const sortedLines = [...new Set(lines)].sort((a, b) => a - b);
  const ranges: string[] = [];
  let start = sortedLines[0];
  let end = sortedLines[0];
  
  for (let i = 1; i < sortedLines.length; i++) {
    if (sortedLines[i] === end + 1) {
      end = sortedLines[i];
    } else {
      if (start === end) {
        ranges.push(start.toString());
      } else {
        ranges.push(`${start}-${end}`);
      }
      start = end = sortedLines[i];
    }
  }
  
  // Add final range
  if (start === end) {
    ranges.push(start.toString());
  } else {
    ranges.push(`${start}-${end}`);
  }
  
  return ranges.join(',');
};

/**
 * Generate a claude-was-here note in the standard format
 */
const generateClaudeNote = (claudeLineMapping: ClaudeLineMapping, contentSignatures?: ContentSignatures): string => {
  let output = 'claude-was-here\nversion: 1.1\n';
  
  const filesWithLines = Object.keys(claudeLineMapping).filter(
    filepath => claudeLineMapping[filepath].size > 0
  );
  
  if (filesWithLines.length > 0) {
    const maxLength = Math.max(...filesWithLines.map(path => path.length));
    
    for (const filepath of filesWithLines.sort()) {
      const lineSet = claudeLineMapping[filepath];
      const ranges = convertLinesToRanges(Array.from(lineSet));
      if (ranges) {
        const paddedPath = `${filepath}:`.padEnd(maxLength + 2);
        output += `${paddedPath} ${ranges}\n`;
      }
    }
  }
  
  // Add content signatures if present
  if (contentSignatures && contentSignatures.hashes.size > 0) {
    output += '\n'; // Empty line separator
    output += `content-signatures: ${Array.from(contentSignatures.hashes).join(',')}\n`;
  }
  
  return output;
};

/**
 * Main analysis function that combines all steps
 */
const analyzePRSquashClaudeContributions = async (
  testDir: string,
  baseCommit: string,
  headCommit: string
): Promise<string> => {
  // Step 1: Collect Claude notes from all commits in the PR
  const { contributions, contentSignatures } = await collectClaudeNotesFromCommits(testDir, baseCommit, headCommit);
  
  // Step 2: Consolidate Claude contributions directly from git notes data
  const claudeLineMapping = await consolidateClaudeContributions(testDir, contributions);
  
  // Step 3: Generate the final note with preserved content signatures
  return generateClaudeNote(claudeLineMapping, contentSignatures);
};

async function main() {
  if (process.argv.length < 5) {
    console.error("Usage: bun run analyze-claude-lines.ts <claude_data_file> <base_commit> <latest_commit>");
    console.error("");
    console.error("This script analyzes Claude Code contributions across commits and maps them");
    console.error("to the final diff, generating accurate attribution for squashed commits.");
    process.exit(1);
  }
  
  const claudeDataFile = process.argv[2];
  const baseCommit = process.argv[3];
  const latestCommit = process.argv[4];
  
  try {
    // Verify the Claude data file exists
    await readFile(claudeDataFile, 'utf-8');
    console.error(`Processing Claude contributions from ${claudeDataFile}`);
    
    // Use the embedded analysis logic
    const noteContent = await analyzePRSquashClaudeContributions(
      process.cwd(),
      baseCommit, 
      latestCommit
    );
    
    // Output the final note content
    console.log(noteContent);
    
  } catch (error) {
    console.error('Error analyzing Claude contributions:', error);
    process.exit(1);
  }
}

main().catch(console.error);